File: Python3/Trie/208. Implement Trie (Prefix Tree).py
""" 208. Implement Trie (Prefix Tree)
Difficulty: Medium
Submisssion Link: https://leetcode.com/problems/implement-trie-prefix-tree/submissions/1387456951/ """

class TrieNode:
    def __init__(self):
        self.child = [None] * 26
        self.end = False
class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        newNode = self.root
        for char in word:
            if newNode.child[ord(char) - ord('a')] == None:
                newNode.child[ord(char) - ord('a')] = TrieNode()
            newNode = newNode.child[ord(char) - ord('a')]
        newNode.end = True

    def search(self, word: str) -> bool:
        newNode = self.root
        for char in word:
            if newNode.child[ord(char) - ord('a')] == None:
                return False
            newNode = newNode.child[ord(char) - ord('a')]

        return True if newNode.end == True else False

    def startsWith(self, prefix: str) -> bool:
        newNode = self.root
        for char in prefix:
            if newNode.child[ord(char) - ord('a')] == None:
                return False
            newNode = newNode.child[ord(char) - ord('a')]
        return True

File: C#/Trie/208. Implement Trie (Prefix Tree).cs
/* 208. Implement Trie (Prefix Tree)
Difficulty: Medium
Submisssion Link: https://leetcode.com/problems/implement-trie-prefix-tree/submissions/1387457625/ */

public class Trie
{
  HashSet<string> values;
  HashSet<string> keys;
  public Trie()
  {
    values = new HashSet<string>();
    keys = new HashSet<string>();
  }
    
  public void Insert(string word)
  {
    if (!values.Contains(word))
      values.Add(word);
    for (var i = word.Length; i > 0; i--)
    {
      if (!keys.Contains(word.Substring(0,i)))
        keys.Add(word.Substring(0,i));
      else
        break;
    }
  }
    
  public bool Search(string word)
  {
    return values.Contains(word);
  }
    
  public bool StartsWith(string prefix)
  {
    return keys.Contains(prefix);
  }
}